File: app/src/main/java/br/com/pedrosilva/tecnonutri/presentation/ui/adapters/FeedAdapter.java
package br.com.pedrosilva.tecnonutri.presentation.ui.adapters;

import android.content.Context;
import android.content.res.Resources;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import br.com.pedrosilva.tecnonutri.R;
import br.com.pedrosilva.tecnonutri.domain.entities.FeedItem;
import br.com.pedrosilva.tecnonutri.domain.entities.Profile;
import br.com.pedrosilva.tecnonutri.presentation.ui.components.CircleTransformation;
import br.com.pedrosilva.tecnonutri.presentation.ui.listeners.ChangeLikeListener;
import br.com.pedrosilva.tecnonutri.presentation.ui.listeners.FeedItemClickListener;
import br.com.pedrosilva.tecnonutri.presentation.ui.listeners.ProfileClickListener;
import br.com.pedrosilva.tecnonutri.presentation.ui.listeners.SingleOnClickListener;
import br.com.pedrosilva.tecnonutri.util.AppUtil;

/**
 * Created by psilva on 3/16/17.
 */

public class FeedAdapter extends GenericAdapter<FeedItem, RecyclerView.ViewHolder> {

    private FeedItemClickListener feedItemClickListener;
    private ProfileClickListener profileClickListener;
    private ChangeLikeListener changeLikeListener;

    public FeedAdapter(Context context) {
        super(context, true);
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        RecyclerView.ViewHolder viewHolder;
        if (viewType == VIEWTYPE_ITEM) {
            View view = inflater.inflate(R.layout.item_feed, parent, false);
            viewHolder = new ViewHolder(
                    view,
                    feedItemClickListener,
                    profileClickListener,
                    changeLikeListener
            );
        } else {
            View view = inflater.inflate(R.layout.progress, parent, false);
            viewHolder = new ViewHolderLoader(view);
        }
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder.getItemViewType() == VIEWTYPE_ITEM) {
            final FeedItem item = getItem(position);
            ((ViewHolder) holder).setup(item);
        } else {
            ((ViewHolderLoader) holder).setup();
        }
    }

    public void setFeedItemClickListener(FeedItemClickListener feedItemClickListener) {
        this.feedItemClickListener = feedItemClickListener;
    }

    public void setProfileClickListener(ProfileClickListener profileClickListener) {
        this.profileClickListener = profileClickListener;
    }

    public void setChangeLikeListener(ChangeLikeListener changeLikeListener) {
        this.changeLikeListener = changeLikeListener;
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        private final FeedItemClickListener feedItemClickListener;
        private final ProfileClickListener profileClickListener;
        private final ChangeLikeListener changeLikeListener;

        private TextView tvProfileName;
        private ImageView ivProfileImage;
        private TextView tvProfileGeneralGoal;
        private ImageView ivMeal;
        private ProgressBar pbLoadingImage;
        private TextView tvMealDate;
        private TextView tvEnergy;
        private RelativeLayout rlHeader;
        private CheckBox cbLike;

        ViewHolder(final View v, FeedItemClickListener feedItemClickListener, ProfileClickListener profileClickListener, ChangeLikeListener changeLikeListener) {
            super(v);

            this.feedItemClickListener = feedItemClickListener;
            this.profileClickListener = profileClickListener;
            this.changeLikeListener = changeLikeListener;

            bindElements(v);
            v.setOnClickListener(new FeedItemClick(feedItemClickListener, this));
        }

        private void bindElements(View v) {
            tvProfileName = (TextView) v.findViewById(R.id.tv_profile_name);
            ivProfileImage = (ImageView) v.findViewById(R.id.iv_profile_image);
            tvProfileGeneralGoal = (TextView) v.findViewById(R.id.tv_profile_general_goal);
            ivMeal = (ImageView) v.findViewById(R.id.iv_meal);
            pbLoadingImage = (ProgressBar) v.findViewById(R.id.pb_loading_image);
            tvMealDate = (TextView) v.findViewById(R.id.tv_meal_date);
            tvEnergy = (TextView) v.findViewById(R.id.tv_energy);
            rlHeader = (RelativeLayout) v.findViewById(R.id.rl_header);
            cbLike = (CheckBox) v.findViewById(R.id.cb_like);
        }

        public void setup(final FeedItem feedItem) {
            final Context context = itemView.getContext();
            final Resources resources = context.getResources();

            cbLike.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(CompoundButton compoundButton, boolean checked) {
                    if (feedItem != null && changeLikeListener != null) {
                        changeLikeListener.onChange(feedItem.getFeedHash(), checked);
                    }
                }
            });
            rlHeader.setOnClickListener(new ProfileClick(profileClickListener, this));

            ivProfileImage.setImageResource(R.drawable.profile_image_placeholder);
            ivMeal.setImageDrawable(null);
            pbLoadingImage.setVisibility(View.VISIBLE);

            Profile profile = feedItem.getProfile();

            cbLike.setChecked(feedItem.isLiked());

            tvProfileName.setText(profile.getName());
            tvProfileGeneralGoal.setText(profile.getGeneralGoal());

            String dateFormatted = AppUtil.formatDate(feedItem.getDate());
            String mealDate = resources.getString(R.string.meal_date, dateFormatted);
            tvMealDate.setText(mealDate);

            String kcal = resources.getString(R.string.qty_energy, AppUtil.formatKcal(feedItem.getEnergy()));
            tvEnergy.setText(kcal);

            Picasso.with(context)
                    .load(profile.getImageUrl())
                    .placeholder(R.drawable.profile_image_placeholder)
                    .error(R.drawable.profile_image_placeholder)
                    .transform(new CircleTransformation())
                    .fit()
                    .centerCrop()
                    .into(ivProfileImage);

            Picasso.with(context)
                    .load(feedItem.getImageUrl())
                    .fit()
                    .centerCrop()
                    .into(ivMeal, new Callback() {
                        @Override
                        public void onSuccess() {
                            pbLoadingImage.setVisibility(View.GONE);
                        }

                        @Override
                        public void onError() {
                        }
                    });
        }
    }

    private class FeedItemClick extends SingleOnClickListener {

        private final ViewHolder vh;
        private final FeedItemClickListener listener;

        FeedItemClick(FeedItemClickListener listener, ViewHolder vh) {
            this.listener = listener;
            this.vh = vh;
        }

        @Override
        public void onItemClick(View view) {
            if (listener != null) {
                FeedItem feedItem = getItem(vh.getAdapterPosition());
                if (feedItem != null) {
                    listener.onClickView(feedItem);
                }
            }
        }
    }

    private class ProfileClick extends SingleOnClickListener {

        private final ViewHolder vh;
        private final ProfileClickListener listener;

        ProfileClick(ProfileClickListener listener, ViewHolder vh) {
            this.listener = listener;
            this.vh = vh;
        }

        @Override
        public void onItemClick(View view) {
            if (listener != null) {
                FeedItem feedItem = getItem(vh.getAdapterPosition());
                if (feedItem != null) {
                    Profile profile = feedItem.getProfile();
                    listener.onClickView(profile);
                }
            }
        }
    }
}

File: app/src/main/kotlin/br/com/pedrosilva/tecnonutri/presentation/ui/adapters/FeedAdapter.kt
package br.com.pedrosilva.tecnonutri.presentation.ui.adapters

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.RelativeLayout
import android.widget.TextView
import br.com.pedrosilva.tecnonutri.R
import br.com.pedrosilva.tecnonutri.domain.entities.FeedItem
import br.com.pedrosilva.tecnonutri.domain.entities.Profile
import br.com.pedrosilva.tecnonutri.presentation.ui.components.CircleTransformation
import br.com.pedrosilva.tecnonutri.presentation.ui.listeners.ChangeLikeListener
import br.com.pedrosilva.tecnonutri.presentation.ui.listeners.FeedItemClickListener
import br.com.pedrosilva.tecnonutri.presentation.ui.listeners.ProfileClickListener
import br.com.pedrosilva.tecnonutri.presentation.ui.listeners.setSingleOnClickListener
import br.com.pedrosilva.tecnonutri.util.AppUtil
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso

class FeedAdapter : GenericAdapter<FeedItem, RecyclerView.ViewHolder>(true) {

    private var feedItemClickListener: FeedItemClickListener? = null
    private var profileClickListener: ProfileClickListener? = null
    private var changeLikeListener: ChangeLikeListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return if (viewType == VIEW_TYPE_ITEM) {
            val view = inflater.inflate(R.layout.item_feed, parent, false)
            ViewHolder(
                view,
                feedItemClickListener,
                profileClickListener,
                changeLikeListener
            )
        } else {
            val view = inflater.inflate(R.layout.progress, parent, false)
            ViewHolderLoader(view)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder.itemViewType == VIEW_TYPE_ITEM) {
            val item = getItem(position)
            (holder as? ViewHolder)?.bindElements(item)
        } else {
            (holder as? GenericAdapter<*, *>.ViewHolderLoader)?.setup()
        }
    }

    fun setFeedItemClickListener(feedItemClickListener: FeedItemClickListener?) {
        this.feedItemClickListener = feedItemClickListener
    }

    fun setProfileClickListener(profileClickListener: ProfileClickListener?) {
        this.profileClickListener = profileClickListener
    }

    fun setChangeLikeListener(changeLikeListener: ChangeLikeListener?) {
        this.changeLikeListener = changeLikeListener
    }

    internal inner class ViewHolder(
        view: View,
        private val feedItemClickListener: FeedItemClickListener?,
        private val profileClickListener: ProfileClickListener?,
        private val changeLikeListener: ChangeLikeListener?
    ) : RecyclerView.ViewHolder(view) {

        private val tvProfileName: TextView = view.findViewById(R.id.tv_profile_name)
        private val ivProfileImage: ImageView = view.findViewById(R.id.iv_profile_image)
        private val tvProfileGeneralGoal: TextView = view.findViewById(R.id.tv_profile_general_goal)
        private val ivMeal: ImageView = view.findViewById(R.id.iv_meal)
        private val pbLoadingImage: ProgressBar = view.findViewById(R.id.pb_loading_image)
        private val tvMealDate: TextView = view.findViewById(R.id.tv_meal_date)
        private val tvEnergy: TextView = view.findViewById(R.id.tv_energy)
        private val rlHeader: RelativeLayout = view.findViewById(R.id.rl_header)
        private val cbLike: CheckBox = view.findViewById(R.id.cb_like)

        fun bindElements(feedItem: FeedItem) {
            bindEvents(feedItem)

            val profile = feedItem.profile
            val resources = itemView.context.resources
            val dateFormatted = AppUtil.formatDate(feedItem.date)

            ivProfileImage.setImageResource(R.drawable.profile_image_placeholder)
            ivMeal.setImageDrawable(null)
            pbLoadingImage.visibility = View.VISIBLE
            cbLike.isChecked = feedItem.isLiked
            tvProfileName.text = profile.name
            tvProfileGeneralGoal.text = profile.generalGoal
            tvMealDate.text = resources.getString(R.string.meal_date, dateFormatted)
            tvEnergy.text = resources.getString(R.string.qty_energy, AppUtil.formatKcal(feedItem.energy!!))

            loadImages(profile, feedItem)
        }

        private fun bindEvents(feedItem: FeedItem) {
            cbLike.setOnCheckedChangeListener { _, checked ->
                changeLikeListener?.invoke(feedItem.id, checked)
            }
            itemView.setSingleOnClickListener {
                feedItemClickListener?.invoke(feedItem)
            }
            rlHeader.setSingleOnClickListener {
                profileClickListener?.invoke(feedItem.profile)
            }
        }

        private fun loadImages(profile: Profile, feedItem: FeedItem) {
            Picasso.get()
                .load(profile.imageUrl)
                .placeholder(R.drawable.profile_image_placeholder)
                .error(R.drawable.profile_image_placeholder)
                .transform(CircleTransformation())
                .fit()
                .centerCrop()
                .into(ivProfileImage)

            Picasso.get()
                .load(feedItem.imageUrl)
                .fit()
                .centerCrop()
                .into(ivMeal, object : Callback {
                    override fun onSuccess() {
                        pbLoadingImage.visibility = View.GONE
                    }

                    override fun onError(e: Exception) {
                        // do nothing
                    }
                })
        }
    }
}